packages/maps/src/index.ts

```diff
@@ -1,54 +1,147 @@
+    };
+
+    this._auth = this.encodeParameter('api_key', this._conf.apiKey);
+    this._client = this.encodeParameter('client', client || DEFAULT_CLIENT_ID);
   }
 
-  public set username(value: string) {
-    this._username = value;
+  public get username(): string {
+    return this._conf.username;
   }
 
   public get apiKey(): string {
-    return this._apiKey;
+    return this._conf.apiKey;
   }
 
-  public set apiKey(value: string) {
-    this._apiKey = value;
+  public get serverURL(): string {
+    return this._conf.serverUrlTemplate.replace('{user}', this._conf.username);
   }
 
-  public instantiateMap(options: { sql?: string, dataset?: string }) {
+  /**
+   * Instantiate a map based on dataset name or sql query, returning a layergroup
+   *
+   * Note: Get layergroup.metadata.tilejson.vector.tiles to consume MVT tiles
+   * @param options
+   */
+  public async instantiateSimpleMap(options: { sql?: string, dataset?: string }) {
     const { sql, dataset } = options;
 
     if (sql === undefined && dataset === undefined) {
       throw new Error('Please provide a dataset or a SQL query');
     }
 
-    const mapParameters = {
+    const mapConfig = {
       layers: [{
         type: 'cartodb',
         options: {
-            sql: sql || `select * from ${dataset}`
+          sql: sql || `select * from ${dataset}`
         }
       }],
       version: '1.3.1'
     };
 
-    const requestOptions = {
+    const layergroup = await this.instantiateMap(mapConfig);
+    return layergroup;
+  }
+
+  private async instantiateMap(mapConfig: any) {
+    let response;
+    try {
+      const payload = JSON.stringify(mapConfig);
+      response = await fetch(this.makeHttpRequest(payload));
+    } catch (error) {
+      throw new Error(`Failed to connect to Maps API with the user ('${this._conf.username}'): ${error}`);
+    }
+
+    const layergroup = await response.json();
+    if (!response.ok) {
+      this.dealWithWindshaftErrors(response, layergroup);
+    }
+
+    return layergroup;
+  }
+
+  private makeHttpRequest(payload: any) {
+    const parameters = [this._conf.auth, this._client, this.encodeParameter('config', payload)];
+    const url = this.generateUrl(this.generateMapsApiUrl(), parameters);
+    if (url.length < REQUEST_GET_MAX_URL_LENGTH) {
+      return this.getRequest(url);
+    }
+
+    return this.postRequest(payload);
+  }
+
+  private dealWithWindshaftErrors(response: any, layergroup: any) {
+    if (response.status === 401) {
+      throw new Error(
+        `Unauthorized access to Maps API: invalid combination of user ('${this._conf._username}') and apiKey ('${this._conf.apiKey}')`
+      );
+    } else if (response.status === 403) {
+      throw new Error(
+        `Unauthorized access to dataset: the provided apiKey('${this._conf.apiKey}') doesn't provide access to the requested data`
+      );
+    }
+    throw new Error(`${JSON.stringify(layergroup.errors)}`);
+  }
+
+  private getRequest(url: string) {
```

Comment:
Those functions (`getRequest`, `postRequest`, `encodeParameter`, `generateUrl`) are generic and we should not add them to the class prototype as a matter of not adding more functions that are not useful for the ones who use the library. We can define them in another file and them import it into this one, or just move it outside the class definition.

Reply:
I aggree. That would be a first step. Maybe we could reuse some of them among different packages a bit later

Reply:
Refactored into a little utils.ts module